### DISCUSSION TITLE
OptionalImportError: from einops import rearrange (No module named 'einops').

### DISCUSSION THREAD

**Discussion by Leeyeonju1** [Project-MONAI/MONAI#5159] [Q&A]
Hi!
I've got this issue while using monai  SWIN-UNETR.
I hope there are some ways to solve this error! 

Thanks!

**Comment by wyli** [comment]
have you tried `pip install "monai[einops]"`?

https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

**Reply by Leeyeonju1** [reply]
Yes, I successfully installed it and  imported it, but still I am getting an error!

**Reply by wyli** [reply]
probably you don't have the latest einops, to upgrade: `pip install -U einops`

**Reply by Oier-Mk** [reply]
not working, the error persists

**Reply by Leeyeonju1** [reply]
I resolved the error. 
I'm currently using google colab. I'm still not sure about what resolved the error.
Maybe you can try pip install monai > pip install einops and import those. 
Also, I'm not importing swin unetr from monai. I copied and pasted the source code in py file and import the py file.
I hope it works for you too.

**Reply by toufiqmusah** [reply]
`pip install -q "monai-weekly[einops]`
This worked for me. Don't forget to restart the runtime if in colab.

**Comment by monajalal** [comment]
(oneposeplus) mona@ard-gpu-01:~/OnePose_Plus_Plus/submodules/LoFTR/demo$ ./run_demo.sh 
Running LoFTR demo..
Traceback (most recent call last):
  File "/home/mona/OnePose_Plus_Plus/submodules/LoFTR/demo/demo_loftr.py", line 19, in <module>
    from src.loftr import LoFTR, default_cfg
  File "/home/mona/OnePose_Plus_Plus/submodules/LoFTR/demo/../src/loftr/__init__.py", line 1, in <module>
    from .loftr import LoFTR
  File "/home/mona/OnePose_Plus_Plus/submodules/LoFTR/demo/../src/loftr/loftr.py", line 3, in <module>
    from einops.einops import rearrange
ModuleNotFoundError: No module named 'einops'

**Comment by adit-negi** [comment]
`pip install "monai[einops]"`
Tried the above out, still the same error as OP persists.

**Reply by KumoLiu** [reply]
Hi @adit-negi, could you please try `pip show einops` to show whether you installed einops successfully?

**Reply by jonathanquantumalpha** [reply]
Hi @KumoLiu I got this output from running pip show einops:

Name: einops
Version: 0.7.0
Summary: A new flavour of deep learning operations
Home-page: 
Author: Alex Rogozhnikov
Author-email: 
License: MIT
Location: /usr/local/lib/python3.10/dist-packages
Requires: 
Required-by: 

But when running the trainer I still got this error in Colab:

---------------------------------------------------------------------------
OptionalImportError                       Traceback (most recent call last)
[<ipython-input-28-8749e73d37e2>](https://localhost:8080/#) in <cell line: 11>()
      9     loss_epochs,
     10     trains_epoch,
---> 11 ) = trainer(
     12     model=model,
     13     train_loader=train_loader,

54 frames
    [... skipping hidden 1 frame]

    [... skipping hidden 1 frame]

    [... skipping hidden 1 frame]

    [... skipping hidden 1 frame]

[/usr/local/lib/python3.10/dist-packages/monai/utils/module.py](https://localhost:8080/#) in optional_import(module, version, version_checker, name, descriptor, version_args, allow_namespace_pkg, as_type)
    394         actual_cmd = f"import {module}"
    395     try:
--> 396         pkg = __import__(module)  # top level module
    397         the_module = import_module(module)
    398         if not allow_namespace_pkg:

OptionalImportError: from einops import rearrange (No module named 'einops').

For details about installing the optional dependencies, please visit:
    https://docs.monai.io/en/latest/installation.html#installing-the-recommended-dependencies

**Reply by KumoLiu** [reply]
Make sure you are in the same environment, such as install and show the einops in the Colab.

**Reply by Navoditamathur** [reply]
Facing the same error
